File: modulo-java/src/com/fabrica/muebles/util/CrearTablasAutomatico.java
package com.fabrica.muebles.util;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.Statement;
import java.sql.SQLException;

/**
 * Programa para crear automáticamente la base de datos y las tablas
 * Sistema de Gestión - Fábrica de Muebles
 * 
 * Tablas: clientes, produccion, proveedor
 * 
 * @author William Alonso Samaca Lopez
 * @version 2.0
 */
public class CrearTablasAutomatico {
    
    // Datos de conexión (sin base de datos específica para crearla primero)
    private static final String URL_SERVER = "jdbc:mysql://localhost:3306/?useSSL=false&serverTimezone=UTC&allowPublicKeyRetrieval=true";
    private static final String USUARIO = "root";
    private static final String CLAVE = "Mariana28";
    
    public static void main(String[] args) {
        System.out.println("════════════════════════════════════════════════════════");
        System.out.println("  CREACIÓN AUTOMÁTICA DE BASE DE DATOS Y TABLAS        ");
        System.out.println("  Sistema: Fábrica de Muebles                          ");
        System.out.println("════════════════════════════════════════════════════════");
        System.out.println();
        
        Connection conn = null;
        Statement stmt = null;
        
        try {
            // PASO 1: Registrar el driver
            System.out.println("Paso 1: Registrando driver JDBC...");
            Class.forName("com.mysql.cj.jdbc.Driver");
            System.out.println("✅ EXITOSO: Driver registrado correctamente");
            System.out.println();
            
            // PASO 2: Conectar al servidor MySQL (sin base de datos específica)
            System.out.println("Paso 2: Conectando al servidor MySQL...");
            conn = DriverManager.getConnection(URL_SERVER, USUARIO, CLAVE);
            stmt = conn.createStatement();
            System.out.println("✅ EXITOSO: Conexión establecida");
            System.out.println();
            
            // PASO 3: Crear la base de datos
            System.out.println("Paso 3: Creando base de datos 'fabrica_muebles'...");
            stmt.executeUpdate("CREATE DATABASE IF NOT EXISTS fabrica_muebles " +
                             "CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci");
            System.out.println("✅ EXITOSO: Base de datos creada");
            System.out.println();
            
            // PASO 4: Usar la base de datos
            stmt.executeUpdate("USE fabrica_muebles");
            
            // ═══════════════════════════════════════════════════════════════
            // PASO 5: Crear tabla CLIENTES
            // ═══════════════════════════════════════════════════════════════
            System.out.println("Paso 4: Creando tabla 'clientes'...");
            stmt.executeUpdate("DROP TABLE IF EXISTS clientes");
            
            String sqlClientes = 
                "CREATE TABLE clientes (" +
                "  id_cliente INT AUTO_INCREMENT PRIMARY KEY," +
                "  nombre VARCHAR(100) NOT NULL," +
                "  telefono VARCHAR(20) NOT NULL," +
                "  direccion VARCHAR(200)," +
                "  email VARCHAR(100) NOT NULL," +
                "  fecha_registro DATE NOT NULL," +
                "  estado INT DEFAULT 1," +
                "  INDEX idx_nombre (nombre)," +
                "  INDEX idx_email (email)" +
                ") ENGINE=InnoDB DEFAULT CHARSET=utf8mb4";
            
            stmt.executeUpdate(sqlClientes);
            System.out.println("✅ EXITOSO: Tabla 'clientes' creada");
            System.out.println();
            
            // ═══════════════════════════════════════════════════════════════
            // PASO 6: Crear tabla PRODUCCION
            // ═══════════════════════════════════════════════════════════════
            System.out.println("Paso 5: Creando tabla 'produccion'...");
            stmt.executeUpdate("DROP TABLE IF EXISTS produccion");
            
            String sqlProduccion = 
                "CREATE TABLE produccion (" +
                "  id INT AUTO_INCREMENT PRIMARY KEY," +
                "  nombre_producto VARCHAR(100) NOT NULL," +
                "  cantidad INT NOT NULL DEFAULT 0," +
                "  fecha_inicio DATE NOT NULL," +
                "  fecha_fin DATE," +
                "  estado VARCHAR(50) NOT NULL," +
                "  INDEX idx_estado (estado)," +
                "  INDEX idx_fecha (fecha_inicio)" +
                ") ENGINE=InnoDB DEFAULT CHARSET=utf8mb4";
            
            stmt.executeUpdate(sqlProduccion);
            System.out.println("✅ EXITOSO: Tabla 'produccion' creada");
            System.out.println();
            
            // ═══════════════════════════════════════════════════════════════
            // PASO 7: Crear tabla PROVEEDOR
            // ═══════════════════════════════════════════════════════════════
            System.out.println("Paso 6: Creando tabla 'proveedor'...");
            stmt.executeUpdate("DROP TABLE IF EXISTS proveedor");
            
            String sqlProveedor = 
                "CREATE TABLE proveedor (" +
                "  id INT AUTO_INCREMENT PRIMARY KEY," +
                "  nombre VARCHAR(100) NOT NULL," +
                "  contacto VARCHAR(100) NOT NULL," +
                "  telefono VARCHAR(20) NOT NULL," +
                "  direccion VARCHAR(200)," +
                "  correo VARCHAR(100) NOT NULL," +
                "  INDEX idx_nombre (nombre)" +
                ") ENGINE=InnoDB DEFAULT CHARSET=utf8mb4";
            
            stmt.executeUpdate(sqlProveedor);
            System.out.println("✅ EXITOSO: Tabla 'proveedor' creada");
            System.out.println();
            
            // ═══════════════════════════════════════════════════════════════
            // PASO 8: Insertar datos de prueba en CLIENTES
            // ═══════════════════════════════════════════════════════════════
            System.out.println("Paso 7: Insertando datos de prueba en 'clientes'...");
            
            String[] clientesInsert = {
                "INSERT INTO clientes (nombre, telefono, direccion, email, fecha_registro, estado) VALUES " +
                "('Juan Pérez García', '3101234567', 'Calle 10 #20-30, Tunja', 'juan.perez@email.com', CURDATE(), 1)",
                
                "INSERT INTO clientes (nombre, telefono, direccion, email, fecha_registro, estado) VALUES " +
                "('María López Rodríguez', '3209876543', 'Carrera 15 #25-40, Tunja', 'maria.lopez@email.com', CURDATE(), 1)",
                
                "INSERT INTO clientes (nombre, telefono, direccion, email, fecha_registro, estado) VALUES " +
                "('Carlos Martínez Silva', '3158765432', 'Avenida Norte #30-15, Tunja', 'carlos.martinez@email.com', CURDATE(), 1)",
                
                "INSERT INTO clientes (nombre, telefono, direccion, email, fecha_registro, estado) VALUES " +
                "('Ana Gómez Torres', '3187654321', 'Calle 5 #12-25, Tunja', 'ana.gomez@email.com', CURDATE(), 1)",
                
                "INSERT INTO clientes (nombre, telefono, direccion, email, fecha_registro, estado) VALUES " +
                "('Luis Hernández Castro', '3145678901', 'Carrera 8 #18-20, Tunja', 'luis.hernandez@email.com', CURDATE(), 1)"
            };
            
            for (String insert : clientesInsert) {
                stmt.executeUpdate(insert);
            }
            System.out.println("✅ EXITOSO: " + clientesInsert.length + " clientes insertados");
            System.out.println();
            
            // ═══════════════════════════════════════════════════════════════
            // PASO 9: Insertar datos de prueba en PRODUCCION
            // ═══════════════════════════════════════════════════════════════
            System.out.println("Paso 8: Insertando datos de prueba en 'produccion'...");
            
            String[] produccionInsert = {
                "INSERT INTO produccion (nombre_producto, cantidad, fecha_inicio, fecha_fin, estado) VALUES " +
                "('Mesa de Comedor Roble', 10, '2024-11-01', '2024-11-15', 'Finalizado')",
                
                "INSERT INTO produccion (nombre_producto, cantidad, fecha_inicio, fecha_fin, estado) VALUES " +
                "('Silla Ejecutiva Ergonómica', 25, '2024-11-10', NULL, 'En Proceso')",
                
                "INSERT INTO produccion (nombre_producto, cantidad, fecha_inicio, fecha_fin, estado) VALUES " +
                "('Escritorio Moderno L-Shape', 8, '2024-11-05', '2024-11-20', 'Finalizado')",
                
                "INSERT INTO produccion (nombre_producto, cantidad, fecha_inicio, fecha_fin, estado) VALUES " +
                "('Estantería de Pared', 15, '2024-11-12', NULL, 'En Proceso')",
                
                "INSERT INTO produccion (nombre_producto, cantidad, fecha_inicio, fecha_fin, estado) VALUES " +
                "('Armario 3 Puertas', 5, '2024-10-25', '2024-11-08', 'Entregado')"
            };
            
            for (String insert : produccionInsert) {
                stmt.executeUpdate(insert);
            }
            System.out.println("✅ EXITOSO: " + produccionInsert.length + " registros de producción insertados");
            System.out.println();
            
            // ═══════════════════════════════════════════════════════════════
            // PASO 10: Insertar datos de prueba en PROVEEDOR
            // ═══════════════════════════════════════════════════════════════
            System.out.println("Paso 9: Insertando datos de prueba en 'proveedor'...");
            
            String[] proveedorInsert = {
                "INSERT INTO proveedor (nombre, contacto, telefono, direccion, correo) VALUES " +
                "('Maderas del Norte S.A.S', 'Pedro Ramírez', '3201234567', 'Zona Industrial Norte, Tunja', 'ventas@maderasnorte.com')",
                
                "INSERT INTO proveedor (nombre, contacto, telefono, direccion, correo) VALUES " +
                "('Herrajes y Accesorios Ltda', 'Sandra Castro', '3159876543', 'Calle 20 #15-30, Tunja', 'pedidos@herrajes.com')",
                
                "INSERT INTO proveedor (nombre, contacto, telefono, direccion, correo) VALUES " +
                "('Pinturas Industriales', 'Miguel Ángel Torres', '3187654321', 'Carrera 10 #25-40, Tunja', 'contacto@pinturas.com')",
                
                "INSERT INTO proveedor (nombre, contacto, telefono, direccion, correo) VALUES " +
                "('Tapizados El Roble', 'Laura Martínez', '3145678901', 'Avenida Sur #18-22, Tunja', 'info@tapizados.com')",
                
                "INSERT INTO proveedor (nombre, contacto, telefono, direccion, correo) VALUES " +
                "('Vidrios y Espejos S.A.', 'Roberto Díaz', '3208765432', 'Zona Comercial Este, Tunja', 'ventas@vidrios.com')"
            };
            
            for (String insert : proveedorInsert) {
                stmt.executeUpdate(insert);
            }
            System.out.println("✅ EXITOSO: " + proveedorInsert.length + " proveedores insertados");
            System.out.println();
            
            // ═══════════════════════════════════════════════════════════════
            // RESUMEN FINAL
            // ═══════════════════════════════════════════════════════════════
            System.out.println("════════════════════════════════════════════════════════");
            System.out.println("    ¡BASE DE DATOS CREADA EXITOSAMENTE!                 ");
            System.out.println("════════════════════════════════════════════════════════");
            System.out.println();
            System.out.println("📊 RESUMEN:");
            System.out.println("   ✅ Base de datos: fabrica_muebles");
            System.out.println("   ✅ Tabla 'clientes' creada con 5 registros");
            System.out.println("   ✅ Tabla 'produccion' creada con 5 registros");
            System.out.println("   ✅ Tabla 'proveedor' creada con 5 registros");
            System.out.println();
            System.out.println("🎉 Tu sistema está listo para usar");
            System.out.println();
            
        } catch (ClassNotFoundException e) {
            System.err.println();
            System.err.println("❌ ERROR: No se encontró el driver JDBC de MySQL");
            System.err.println("   Verifica que mysql-connector-j-9.4.0.jar esté en la carpeta lib/");
            e.printStackTrace();
        } catch (SQLException e) {
            System.err.println();
            System.err.println("❌ ERROR DE SQL: " + e.getMessage());
            System.err.println("   Verifica que MySQL esté ejecutándose");
            System.err.println("   Verifica usuario y contraseña");
            e.printStackTrace();
        } finally {
            // Cerrar recursos
            try {
                if (stmt != null) stmt.close();
                if (conn != null) conn.close();
                System.out.println("🔒 Conexión cerrada");
                System.out.println();
            } catch (SQLException e) {
                e.printStackTrace();
            }
        }
    }
}

File: src/com/fabrica/muebles/util/crear_tablas_automatico.py
import pymysql
from pymysql import Error


class CrearTablasAutomatico:
    """
    Programa para crear automaticamente la base de datos y las tablas
    Sistema de Gestion - Fabrica de Muebles

    Tablas: clientes, produccion, proveedor

    Autor: William Alonso Samaca Lopez
    Version: 2.0
    """

    URL_SERVER = "localhost"
    PORT = 3306
    USUARIO = "root"
    CLAVE = "Mariana28"

    @staticmethod
    def ejecutar():
        print("============================================================")
        print("  CREACION AUTOMATICA DE BASE DE DATOS Y TABLAS")
        print("  Sistema: Fabrica de Muebles")
        print("============================================================\n")

        conexion = None
        cursor = None

        try:
            print("Paso 1: Conectando al servidor MySQL...")
            conexion = pymysql.connect(
                host=CrearTablasAutomatico.URL_SERVER,
                port=CrearTablasAutomatico.PORT,
                user=CrearTablasAutomatico.USUARIO,
                password=CrearTablasAutomatico.CLAVE,
                charset='utf8mb4'
            )
            cursor = conexion.cursor()
            print("EXITOSO: Conexion establecida\n")

            print("Paso 2: Creando base de datos 'fabrica_muebles'...")
            cursor.execute(
                "CREATE DATABASE IF NOT EXISTS fabrica_muebles CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci")
            print("EXITOSO: Base de datos creada\n")

            cursor.execute("USE fabrica_muebles")

            print("Paso 3: Creando tabla 'clientes'...")
            cursor.execute("DROP TABLE IF EXISTS clientes")
            sql_clientes = """
                CREATE TABLE clientes (
                    id_cliente INT AUTO_INCREMENT PRIMARY KEY,
                    nombre VARCHAR(100) NOT NULL,
                    telefono VARCHAR(20) NOT NULL,
                    direccion VARCHAR(200),
                    email VARCHAR(100) NOT NULL,
                    fecha_registro DATE NOT NULL,
                    estado INT DEFAULT 1,
                    INDEX idx_nombre (nombre),
                    INDEX idx_email (email)
                ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4
            """
            cursor.execute(sql_clientes)
            print("EXITOSO: Tabla 'clientes' creada\n")

            print("Paso 4: Creando tabla 'produccion'...")
            cursor.execute("DROP TABLE IF EXISTS produccion")
            sql_produccion = """
                CREATE TABLE produccion (
                    id INT AUTO_INCREMENT PRIMARY KEY,
                    nombre_producto VARCHAR(100) NOT NULL,
                    cantidad INT NOT NULL DEFAULT 0,
                    fecha_inicio DATE NOT NULL,
                    fecha_fin DATE,
                    estado VARCHAR(50) NOT NULL,
                    INDEX idx_estado (estado),
                    INDEX idx_fecha (fecha_inicio)
                ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4
            """
            cursor.execute(sql_produccion)
            print("EXITOSO: Tabla 'produccion' creada\n")

            print("Paso 5: Creando tabla 'proveedor'...")
            cursor.execute("DROP TABLE IF EXISTS proveedor")
            sql_proveedor = """
                CREATE TABLE proveedor (
                    id INT AUTO_INCREMENT PRIMARY KEY,
                    nombre VARCHAR(100) NOT NULL,
                    contacto VARCHAR(100) NOT NULL,
                    telefono VARCHAR(20) NOT NULL,
                    direccion VARCHAR(200),
                    correo VARCHAR(100) NOT NULL,
                    INDEX idx_nombre (nombre)
                ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4
            """
            cursor.execute(sql_proveedor)
            print("EXITOSO: Tabla 'proveedor' creada\n")

            print("Paso 6: Insertando datos de prueba en 'clientes'...")
            clientes_insert = [
                "INSERT INTO clientes (nombre, telefono, direccion, email, fecha_registro, estado) VALUES ('Juan Perez Garcia', '3101234567', 'Calle 10 #20-30, Tunja', 'juan.perez@email.com', CURDATE(), 1)",
                "INSERT INTO clientes (nombre, telefono, direccion, email, fecha_registro, estado) VALUES ('Maria Lopez Rodriguez', '3209876543', 'Carrera 15 #25-40, Tunja', 'maria.lopez@email.com', CURDATE(), 1)",
                "INSERT INTO clientes (nombre, telefono, direccion, email, fecha_registro, estado) VALUES ('Carlos Martinez Silva', '3158765432', 'Avenida Norte #30-15, Tunja', 'carlos.martinez@email.com', CURDATE(), 1)",
                "INSERT INTO clientes (nombre, telefono, direccion, email, fecha_registro, estado) VALUES ('Ana Gomez Torres', '3187654321', 'Calle 5 #12-25, Tunja', 'ana.gomez@email.com', CURDATE(), 1)",
                "INSERT INTO clientes (nombre, telefono, direccion, email, fecha_registro, estado) VALUES ('Luis Hernandez Castro', '3145678901', 'Carrera 8 #18-20, Tunja', 'luis.hernandez@email.com', CURDATE(), 1)"
            ]
            for insert in clientes_insert:
                cursor.execute(insert)
            print(f"EXITOSO: {len(clientes_insert)} clientes insertados\n")

            print("Paso 7: Insertando datos de prueba en 'produccion'...")
            produccion_insert = [
                "INSERT INTO produccion (nombre_producto, cantidad, fecha_inicio, fecha_fin, estado) VALUES ('Mesa de Comedor Roble', 10, '2024-11-01', '2024-11-15', 'Finalizado')",
                "INSERT INTO produccion (nombre_producto, cantidad, fecha_inicio, fecha_fin, estado) VALUES ('Silla Ejecutiva Ergonomica', 25, '2024-11-10', NULL, 'En Proceso')",
                "INSERT INTO produccion (nombre_producto, cantidad, fecha_inicio, fecha_fin, estado) VALUES ('Escritorio Moderno L-Shape', 8, '2024-11-05', '2024-11-20', 'Finalizado')",
                "INSERT INTO produccion (nombre_producto, cantidad, fecha_inicio, fecha_fin, estado) VALUES ('Estanteria de Pared', 15, '2024-11-12', NULL, 'En Proceso')",
                "INSERT INTO produccion (nombre_producto, cantidad, fecha_inicio, fecha_fin, estado) VALUES ('Armario 3 Puertas', 5, '2024-10-25', '2024-11-08', 'Entregado')"
            ]
            for insert in produccion_insert:
                cursor.execute(insert)
            print(f"EXITOSO: {len(produccion_insert)} registros de produccion insertados\n")

            print("Paso 8: Insertando datos de prueba en 'proveedor'...")
            proveedor_insert = [
                "INSERT INTO proveedor (nombre, contacto, telefono, direccion, correo) VALUES ('Maderas del Norte S.A.S', 'Pedro Ramirez', '3201234567', 'Zona Industrial Norte, Tunja', 'ventas@maderasnorte.com')",
                "INSERT INTO proveedor (nombre, contacto, telefono, direccion, correo) VALUES ('Herrajes y Accesorios Ltda', 'Sandra Castro', '3159876543', 'Calle 20 #15-30, Tunja', 'pedidos@herrajes.com')",
                "INSERT INTO proveedor (nombre, contacto, telefono, direccion, correo) VALUES ('Pinturas Industriales', 'Miguel Angel Torres', '3187654321', 'Carrera 10 #25-40, Tunja', 'contacto@pinturas.com')",
                "INSERT INTO proveedor (nombre, contacto, telefono, direccion, correo) VALUES ('Tapizados El Roble', 'Laura Martinez', '3145678901', 'Avenida Sur #18-22, Tunja', 'info@tapizados.com')",
                "INSERT INTO proveedor (nombre, contacto, telefono, direccion, correo) VALUES ('Vidrios y Espejos S.A.', 'Roberto Diaz', '3208765432', 'Zona Comercial Este, Tunja', 'ventas@vidrios.com')"
            ]
            for insert in proveedor_insert:
                cursor.execute(insert)
            print(f"EXITOSO: {len(proveedor_insert)} proveedores insertados\n")

            conexion.commit()

            print("============================================================")
            print("    BASE DE DATOS CREADA EXITOSAMENTE!")
            print("============================================================\n")
            print("RESUMEN:")
            print("   Base de datos: fabrica_muebles")
            print("   Tabla 'clientes' creada con 5 registros")
            print("   Tabla 'produccion' creada con 5 registros")
            print("   Tabla 'proveedor' creada con 5 registros\n")
            print("Tu sistema esta listo para usar\n")

        except Error as e:
            print(f"\nERROR DE SQL: {e}")
            print("Verifica que MySQL este ejecutandose")
            print("Verifica usuario y contrasena")
            if conexion:
                conexion.rollback()
        except Exception as e:
            print(f"\nERROR: {e}")
            if conexion:
                conexion.rollback()
        finally:
            if cursor:
                cursor.close()
            if conexion:
                conexion.close()
                print("Conexion cerrada\n")


if __name__ == "__main__":
    CrearTablasAutomatico.ejecutar()